File: source/dotnet/source/cmf.bus/cmf.eventing.berico/EventRegistration.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using Common.Logging;

using cmf.bus;
using cmf.bus.berico;

namespace cmf.eventing.berico
{
    public class EventRegistration : IRegistration
    {
        protected IEventHandler _handler;
        protected IDictionary<string, string> _registrationInfo;
        protected IEnumerable<IInboundEventProcessor> _inboundChain;
        protected ILog _log;


        public virtual Predicate<Envelope> Filter
        {
            get { return null; }
        }

        public virtual IDictionary<string, string> Info
        {
            get { return _registrationInfo; }
        }


        public EventRegistration(IEventHandler handler, IEnumerable<IInboundEventProcessor> inboundChain)
        {
            _handler = handler;
            _inboundChain = inboundChain;

            _registrationInfo = new Dictionary<string, string>();
            _registrationInfo.SetMessageTopic(handler.Topic);

            _log = LogManager.GetLogger(this.GetType());
        }


        public virtual object Handle(Envelope env)
        {
            _log.Debug("Enter Handle");

            object ev = null;
            object result = null;

            try
            {
                this.ProcessInbound(ref ev, ref env);
                result = _handler.Handle(ev, env.Headers);
            }
            catch (Exception ex)
            {
                _log.Warn("Caught an exception while handling an event", ex);
                result = this.HandleFailed(env, ex);
            }

            _log.Debug("Leave Handle");
            return result;
        }

        public virtual object HandleFailed(Envelope env, Exception ex)
        {
            _log.Debug("Enter HandleFailed");

            // either log & return or log & throw
            try
            {
                _log.Debug("Leave HandleFailed");
                return _handler.HandleFailed(env, ex);
            }
            catch (Exception failedToFail)
            {
                _log.Error("Caught an exception attempting to raise the Failed event", failedToFail);
                throw;
            }
        }


        protected virtual void ProcessInbound(ref object ev, ref Envelope env)
        {
            IDictionary<string, object> processorContext = new Dictionary<string, object>();

            foreach (IInboundEventProcessor processor in _inboundChain)
            {
                processor.ProcessInbound(ref ev, ref env, ref processorContext);
            }
        }
    }
}

File: source/java/src/main/java/cmf/eventing/berico/EventRegistration.java
package cmf.eventing.berico;

import java.lang.reflect.ParameterizedType;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import cmf.bus.Envelope;
import cmf.bus.IEnvelopeFilterPredicate;
import cmf.bus.IRegistration;
import cmf.bus.berico.EnvelopeHeaderConstants;
import cmf.eventing.IEventHandler;
import cmf.eventing.IInboundEventProcessor;

public class EventRegistration implements IRegistration {

	protected IEnvelopeFilterPredicate filterPredicate;
	protected Map<String, String> registrationInfo;
	protected List<IInboundEventProcessor> inboundChain;
	protected IEventHandler eventHandler;
	
	
	@Override
	public IEnvelopeFilterPredicate getFilterPredicate() { return filterPredicate; }
	public void setFilterPredicate(IEnvelopeFilterPredicate filterPredicate) { this.filterPredicate = filterPredicate; }
	
	
	@Override
	public Map<String, String> getRegistrationInfo() { return registrationInfo; }
	public void setRegistrationInfo(Map<String, String> registrationInfo) { this.registrationInfo = registrationInfo; }
	
	
	@SuppressWarnings({ "unchecked", "rawtypes" })
	public EventRegistration(IEventHandler eventHandler, List<IInboundEventProcessor> inboundChain) {
		
		this.eventHandler = eventHandler;
		this.inboundChain = inboundChain;
		
		this.registrationInfo = new HashMap<String, String>();
		this.registrationInfo.put(EnvelopeHeaderConstants.MESSAGE_TOPIC, eventHandler.getEventType().getCanonicalName());
	}
	
	
	@Override
	public Object handle(Envelope env) throws Exception {
		Object ev = null;
		Object result = null;

        if (this.processInbound(ev, env))
        {
            try
            {
                result = eventHandler.handle(ev, env.getHeaders());
            }
            catch (Exception ex)
            {
                result = this.handleFailed(env, ex);
            }
        }

        return result;
	}

	@Override
	public Object handleFailed(Envelope env, Exception ex) throws Exception {
        try
        {
            return eventHandler.handleFailed(env, ex);
        }
        catch (Exception failedToFail)
        {
            throw failedToFail;
        }
	}
	
	
    protected boolean processInbound(Object ev, Envelope env)
    {
    	boolean processed = true;
        Map<String, Object> processorContext = new HashMap<String, Object>();

        try {
            for (IInboundEventProcessor processor : this.inboundChain) {
                if (!processor.processInbound(ev, env, processorContext)) {
                    processed = false;
                    break;
                }
            }
        }
        catch (Exception ex)
        {
            processed = false;
        }

        return processed;
    }
}
